drivers/QSPI.h

```diff
@@ -0,0 +1,227 @@
+     *
+     *  @returns
+     *    Returns QSPI_STATUS_SUCCESS on successful reads and QSPI_STATUS_ERROR on failed reads.
+     */
+    qspi_status_t read(unsigned int instruction, unsigned int address, unsigned int alt, char *rx_buffer, size_t *rx_length);
+    
+    /** Write to QSPI peripheral using custom write instruction, alt values
+     *
+     *  @param instruction Instruction value to be used in instruction phase
+     *  @param address Address to be accessed in QSPI peripheral
+     *  @param alt Alt value to be used in instruction phase
+     *  @param tx_buffer Buffer containing data to be sent to peripheral                          
+     *  @param tx_length Pointer to a variable containing the length of data to be transmitted, and on return this variable will be updated with the actual number of bytes written
+     *
+     *  @returns
+     *    Returns QSPI_STATUS_SUCCESS on successful reads and QSPI_STATUS_ERROR on failed reads.
+     */
+    qspi_status_t write(unsigned int instruction, unsigned int address, unsigned int alt, const char *tx_buffer, size_t *tx_length);
+    
+    /** Perform a transaction to write to an address(a control register) and get the status results
+     *
+     *  @param instruction Instruction value to be used in instruction phase
+     *  @param tx_buffer Buffer containing data to be sent to peripheral                          
+     *  @param tx_length Pointer to a variable containing the length of data to be transmitted, and on return this variable will be updated with the actual number of bytes written
+     *  @param rx_buffer Buffer for data to be read from the peripheral                          
+     *  @param rx_length Pointer to a variable containing the length of rx_buffer, and on return this variable will be updated with the actual number of bytes read
+     *
+     *  @returns
+     *    Returns QSPI_STATUS_SUCCESS on successful reads and QSPI_STATUS_ERROR on failed reads.
+     */
+    qspi_status_t command_transfer(unsigned int instruction, const char *tx_buffer, size_t tx_length, const char *rx_buffer, size_t rx_length);
+        
+    /** Acquire exclusive access to this SPI bus
+     */
+    virtual void lock(void);
+
+    /** Release exclusive access to this SPI bus
+     */
+    virtual void unlock(void);
+
+public:
+    virtual ~QSPI() {
+    }
+
+protected:
+    qspi_t _qspi;
+
+    bool acquire(void);
+    static QSPI *_owner;
+    static SingletonPtr<PlatformMutex> _mutex;
+    qspi_bus_width_t _inst_width; //Bus width for Instruction phase
+    qspi_bus_width_t _address_width; //Bus width for Address phase
+    qspi_address_size_t _address_size;
+    qspi_bus_width_t _alt_width; //Bus width for Alt phase
+    qspi_alt_size_t _alt_size;
+    qspi_bus_width_t _data_width; //Bus width for Data phase
+    qspi_command_t _qspi_command; //QSPI Hal command struct
+    int _num_dummy_cycles; //Number of dummy cycles to be used
+    int _hz; //Bus Frequency
+    int _mode; //SPI mode
```

Comment:
small optimization, but this could be only few modes, I believe just two so 8bit variable shall be sufficient to capture the mode?

Reply:
Well, its better to have int vs char for performance reasons in terms of CPU cycles. But if you are concerned about saving few bytes of memory we can optimize.

Reply:
OK as it is